rust/kernel/init.rs

```diff
@@ -0,0 +1,827 @@
+/// #     b: Bar,
+/// # }
+/// # struct Bar {
+/// #     x: u32,
+/// # }
+/// # impl Foo {
+/// #     fn new() -> impl Init<Self> {
+/// #         init!(Self {
+/// #             a: 42,
+/// #             b: Bar {
+/// #                 x: 64,
+/// #             },
+/// #         })
+/// #     }
+/// # }
+/// struct FooContainer {
+///     foo1: Foo,
+///     foo2: Foo,
+///     other: u32,
+/// }
+///
+/// impl FooContainer {
+///     fn new(other: u32) -> impl Init<Self> {
+///         init!(Self {
+///             foo1: Foo::new(),
+///             foo2: Foo::new(),
+///             other,
+///         })
+///     }
+/// }
+/// ```
+#[macro_export]
+macro_rules! init {
+    ($t:ident $(<$($generics:ty),* $(,)?>)? {
+        $($field:ident $(: $val:expr)?),*
+        $(,)?
+    }) => {{
+        // we do not want to allow arbitrary returns
+        struct __InitOk;
+        let init = move |slot: *mut $t $(<$($generics),*>)?| -> ::core::result::Result<__InitOk, _> {
+            {
+                // shadow the structure so it cannot be used to return early
+                struct __InitOk;
+                $(
+                    $(let $field = $val;)?
+                    // call the initializer
+                    // SAFETY: slot is valid, because we are inside of an initializer closure, we return
+                    //         when an error/panic occurs.
+                    unsafe { $crate::init::__private::__InitImpl::__init($field, ::core::ptr::addr_of_mut!((*slot).$field))? };
+                    // create the drop guard
+                    // SAFETY: we forget the guard later when initialization has succeeded.
+                    let $field = unsafe { $crate::init::__private::DropGuard::new(::core::ptr::addr_of_mut!((*slot).$field)) };
+                )*
+                #[allow(unreachable_code, clippy::diverging_sub_expression)]
+                if false {
+                    let _: $t $(<$($generics),*>)? = $t {
+                        $($field: ::core::todo!()),*
+                    };
+                }
+                $(
+                    // forget each guard
+                    ::core::mem::forget($field);
+                )*
+            }
+            Ok(__InitOk)
+        };
+        let init = move |slot: *mut $t $(<$($generics),*>)?| -> ::core::result::Result<(), _> {
+            init(slot).map(|__InitOk| ())
+        };
+        let init: $crate::init::InitClosure<_, $t $(<$($generics),*>)?, _> = unsafe { $crate::init::InitClosure::from_closure(init) };
+        init
+    }}
+}
+
+/// Used to specify the pin information of the fields of a struct.
+///
+/// This is somewhat similar in purpose as
+/// [pin-project-lite](https://crates.io/crates/pin-project-lite).
+/// Place this macro around a struct definition and then `#[pin]` in front of the attributes of each
+/// field you want to have structurally pinned.
+///
+/// needed to use `pin_init`.
+#[macro_export]
+macro_rules! pin_data {
```

Comment:
I would probably mention `#[pin_data]` here if that is what users will normally use.

Actually, is there some case where users would need `pin_data!` directly? If not, should we make this `#[doc(hidden)]`?

Reply:
not that I can think of... so we should hide it